### PR TITLE
Optimize the sync batchnorm by batching the communication

### PR DESCRIPTION
**Pull request by lly-zero-one**
In this PR, we mainly tried to optimize the performance of Syncatchnorm and also fixed one potential issue in the welford_parallel kernel implementation.
1) For performance improvement, we batched the mean/var/count all_gather communication together and sent it once in the forward path
2) We also batch the all_reduce in backward path
3) We add the contiguous call on the input of welford_parallel kernel.

If there is any standard perf benchmark, I would be happy to run it.

### CODE REVIEW

**Comment by lly-zero-one** [comment]
cc @ngimel

**Review by jjsjann123 — APPROVED** [review]
Don't see any major issues and it's passing local test. I'm merging this.

**Comment by Timen** [comment]
The changes introduced in this PR actually reduced my training speed by about 30%, I'm using a 4 GPU setup and O1 fp16 optimizations. Not sure why but I'm reverting to the previous commit.

**Comment by lly-zero-one** [comment]
> The changes introduced in this PR actually reduced my training speed by about 30%, I'm using a 4 GPU setup and O1 fp16 optimizations. Not sure why but I'm reverting to the previous commit.

Can you share the script for reproducing the regression?

**Comment by Timen** [comment]
I'm unfortunately not at liberty to share my training script. I will say however it is a quite complex model setup training on videos and using [NVIDIA/DALI](https://github.com/NVIDIA/DALI) for data loading. I'm wondering what you used to benchmark your changes and where you saw an improvement in training speed?

**Comment by lly-zero-one** [comment]
What if you check the trace and see which part of SyncBatchNorm got slow down? Let me paste the EfficientNet result here.
